src/shell/info.c

```diff
@@ -243,6 +243,49 @@ static int get_per_resource_option (struct jobspec *jobspec,
     return 0;
 }
 
+struct taskmap *create_taskmap (struct shell_info *info)
+{
+    struct taskmap *map = taskmap_create ();
+    if (!map)
+        return NULL;
+    for (int i = 0; i < info->shell_size; i++) {
+        struct rcalc_rankinfo ri;
+        if (rcalc_get_nth (info->rcalc, i, &ri) < 0
+            || taskmap_append (map, i, 1, ri.ntasks) < 0) {
+            shell_log_errno ("taskmap: failed to process rank=%d", i);
+            goto error;
+        }
+    }
+    return map;
+error:
+    taskmap_destroy (map);
+    return NULL;
+}
+
+int shell_info_set_taskmap (struct shell_info *info,
+                            struct taskmap *map)
+{
+    flux_error_t error;
+    const struct idset *taskids;
+    if (!info || !map) {
+        errno = EINVAL;
+        return -1;
+    }
+    if (info->taskmap
+        && taskmap_check (info->taskmap, map, &error) < 0) {
+        shell_log_error ("invalid taskmap: %s", error.text);
+        return -1;
+    }
+    if (!(taskids = taskmap_taskids (map, info->shell_rank)))
+        return -1;
+    idset_destroy (info->taskids);
+    taskmap_destroy (info->taskmap);
+    info->taskmap = map;
+    if (!(info->taskids = idset_copy (taskids)))
+        return -1;
```

Comment:
If `idset_copy()` fails, does the fact that `map` is assigned to the taskmap lead to a double free later when both `shell_info_destroy()` and a caller try to discard the taskmap?

(Actually looking at `shell_taskmap()`, a caller, this does not seem to be the case on this failure path but the taskmap is leaked if `set_taskmap()` fails earlier.)

Anyway, error paths might need auditing there.

Reply:
I will clean up `shell_set_taskmap()` though